daemon/src/GHCSpecter/Control.hs

```diff
@@ -0,0 +1,251 @@
+    HasSourceViewUI (..),
+    HasTimingUI (..),
+    HasUIState (..),
+    ModuleGraphUI (..),
+    UIState (..),
+  )
+import GHCSpecter.UI.Types.Event
+  ( Event (..),
+    ModuleGraphEvent (..),
+    SessionEvent (..),
+    SubModuleEvent (..),
+    TimingEvent (..),
+  )
+import System.IO (IOMode (..), withFile)
+import System.IO.Unsafe (unsafePerformIO)
+
+data ControlF r
+  = GetState ((UIState, ServerState) -> r)
+  | PutState (UIState, ServerState) r
+  | NextEvent (Event -> r)
+  | PrintMsg Text r
+  | GetCurrentTime (UTCTime -> r)
+  | GetLastUpdatedUI (UTCTime -> r)
+  | ShouldUpdate Bool r
+  | SaveSession r
+  deriving (Functor)
+
+type Control = Free ControlF
+
+getState :: Control (UIState, ServerState)
+getState = liftF (GetState id)
+
+putState :: (UIState, ServerState) -> Control ()
+putState (ui, ss) = liftF (PutState (ui, ss) ())
+
+nextEvent :: Control Event
+nextEvent = liftF (NextEvent id)
+
+printMsg :: Text -> Control ()
+printMsg txt = liftF (PrintMsg txt ())
+
+getCurrentTime :: Control UTCTime
+getCurrentTime = liftF (GetCurrentTime id)
+
+getLastUpdatedUI :: Control UTCTime
+getLastUpdatedUI = liftF (GetLastUpdatedUI id)
+
+shouldUpdate :: Bool -> Control ()
+shouldUpdate b = liftF (ShouldUpdate b ())
+
+saveSession :: Control ()
+saveSession = liftF (SaveSession ())
+
+-- TODO: remove this
+tempRef :: IORef Int
+tempRef = unsafePerformIO (newIORef 0)
+{-# NOINLINE tempRef #-}
+
+type Runner = StateT (UIState, ServerState) (Widget IHTML)
+
+-- | step interpretation
+stepControl ::
+  Control r ->
+  Runner
+    ( Either
+        (Control r)
+        ( Either
+            (Event -> Control r)
+            r
+        )
+    )
+stepControl (Pure r) = pure (Right (Right r))
+stepControl (Free (GetState cont)) = do
+  (ui, ss) <- get
+  pure (Left (cont (ui, ss)))
+stepControl (Free (PutState (ui, ss) next)) = do
+  put (ui, ss)
+  pure (Left next)
+stepControl (Free (NextEvent cont)) =
+  pure (Right (Left cont))
+stepControl (Free (PrintMsg txt next)) = do
+  lift $
+    unsafeBlockingIO $ do
+      n <- readIORef tempRef
+      modifyIORef' tempRef (+ 1)
+      TIO.putStrLn $ (T.pack (show n) <> " : " <> txt)
+  pure (Left next)
+stepControl (Free (GetCurrentTime cont)) = do
+  now <- lift $ unsafeBlockingIO Clock.getCurrentTime
+  pure (Left (cont now))
+stepControl (Free (GetLastUpdatedUI cont)) = do
+  lastUpdatedUI <- (^. _1 . uiLastUpdated) <$> get
+  pure (Left (cont lastUpdatedUI))
+stepControl (Free (ShouldUpdate b next)) = do
+  modify' (_1 . uiShouldUpdate .~ b)
+  pure (Left next)
+stepControl (Free (SaveSession next)) = do
+  (_, ss) <- get
+  -- TODO: use asynchronous worker
+  liftIO $
+    withFile "session.json" WriteMode $ \h ->
+      BL.hPutStr h (encode ss)
+  pure (Left next)
+
+stepControlUpToEvent ::
+  Event ->
+  (Event -> Control r) ->
+  Runner (Either (Event -> Control r) r)
+stepControlUpToEvent ev cont0 = loopM stepControl (cont0 ev)
+
+uiUpdateInterval :: NominalDiffTime
```

Comment:
this constants should be located in a separate module, say GHCSpecter.UI.Constants

Reply:
Moved uiUpdateInterval and chanUpdateInterval to that module.